### PR TITLE
Fixup async cancellation

### PR DESCRIPTION
**Pull request by jzila**

### CODE REVIEW

File: codec.go [hunk 1]
```diff
@@ -63,7 +64,7 @@ func newFramedMsgpackEncoder(writer io.Writer) *framedMsgpackEncoder {
 	e := &framedMsgpackEncoder{
 		handle:   newCodecMsgpackHandle(),
 		writer:   writer,
-		writeCh:  make(chan writeBundle),
+		writeCh:  make(chan writeBundle, 128),
```
Comment (akalin-keybase) [inline codec.go]: don't need this anymore, right

File: dispatch.go [hunk 2]
```diff
@@ -40,13 +40,20 @@ func (d *dispatch) Call(ctx context.Context, name string, arg interface{}, res i
 	profiler := d.log.StartProfiler("call %s", name)
 	defer profiler.Stop()
 
+	select {
```
Comment (akalin-keybase) [inline dispatch.go]: Hmm...why is this needed here?

Actually, I'm a bit worried that the body of Encode doesn't wait for <-d.stopCh, too. is doneCh closed when stopCh is?

Comment (jzila) [inline dispatch.go]: They're independent.

File: protocol_test.go [hunk 3]
```diff
@@ -35,6 +39,14 @@ func prepClient(t *testing.T) (TestClient, net.Conn) {
 }
 
 func prepTest(t *testing.T) (TestClient, chan error, net.Conn) {
+	sigs := make(chan os.Signal, 1)
```
Comment (akalin-keybase) [inline protocol_test.go]: remove debug code

File: test_utils.go [hunk 4]
```diff
@@ -361,3 +362,18 @@ func (eu *mockErrorUnwrapper) MakeArg() interface{} {
 func (eu *mockErrorUnwrapper) UnwrapError(i interface{}) (appErr error, dispatchErr error) {
 	return nil, nil
 }
+
+type blockingBuffer struct {
```
Comment (akalin-keybase) [inline test_utils.go]: here, too

File: request.go [hunk 5]
```diff
@@ -60,11 +60,14 @@ func (r *callRequest) Reply(enc encoder, res interface{}, errArg interface{}) (e
 			errArg,
 			res,
 		}
-		errCh := enc.Encode(v)
-		// TODO investigate hypothetical server-side hang here
-		err = <-errCh
-		if err != nil {
-			r.log.Warning("Reply error for %d: %s", r.SeqNo(), err.Error())
+		errCh := enc.Encode(context.Background(), v)
```
Comment (akalin-keybase) [inline request.go]: not r.ctx?

File: request.go [hunk 6]
```diff
@@ -60,11 +60,14 @@ func (r *callRequest) Reply(enc encoder, res interface{}, errArg interface{}) (e
...
+		select {
+		case err := <-errCh:
+			if err != nil {
+				r.log.Warning("Reply error for %d: %s", r.SeqNo(), err.Error())
+			}
+		default:
+			// Don't wait for a reply
```
Comment (akalin-keybase) [inline request.go]: hmm, i'm not so sure about this...why was this needed?

**Comment by akalin-keybase** [comment]
First pass done, while you investigate why KBFS case is still hanging

File: request.go (hunk 6 above)
Comment (jzila) [inline request.go]: Otherwise the goroutine can leak.

File: request.go (hunk 5 above)
Comment (jzila) [inline request.go]: Maybe that's why it's hanging.

File: test_utils.go (hunk 4 above)
Comment (jzila) [inline test_utils.go]: This isn't debug code, it supports a valid test.

File: dispatch.go (hunk 2 above)
Comment (akalin-keybase) [inline dispatch.go]: Well, in that case, why doesn't Encode wait on <-d.stopCh, too?

File: test_utils.go (hunk 4 above)
Comment (akalin-keybase) [inline test_utils.go]: oh...any particular reason why you don't just use net.Pipe?

Comment (jzila) [inline test_utils.go]: Didn't know it existed :P

File: dispatch.go (hunk 2 above)
Comment (jzila) [inline dispatch.go]: They're independent, but they're both closed by the transport. It'll stop correctly as is.

File: codec.go [hunk 7]
```diff
@@ -100,11 +102,40 @@ func (e *framedMsgpackEncoder) Encode(i interface{}) <-chan error {
...
 	case <-e.doneCh:
 		ch <- io.EOF
+	case <-ctx.Done():
+		ch <- ctx.Err()
 	case e.writeCh <- writeBundle{bytes, ch}:
 	}
 	return ch
 }
 
+func (e *framedMsgpackEncoder) EncodeAndWriteAsync(ctx context.Context, i interface{}) <-chan error {
```
Comment (akalin-keybase) [inline codec.go]: The caller of this function just passes in context.Background(), so may as well remove the ctx parameter and the cases below

File: dispatch.go (hunk 2 above)
Comment (akalin-keybase) [inline dispatch.go]: Okay, cool.

File: dispatch.go [hunk 8]
```diff
@@ -40,13 +40,20 @@ func (d *dispatch) Call(ctx context.Context, name string, arg interface{}, res i
...
+	case <-d.stopCh:
```
Comment (akalin-keybase) [inline dispatch.go]: Actually, this isn't needed for correctness, right? Since if we're closed we'll hit the stopCh path below or the errCh path. If so, I'd rather just remove this, as it muddies the waters.

Comment (jzila) [inline dispatch.go]: Yeah but I'd rather not even attempt to encode if we're supposed to be stopped. The key here is that once `Close` is called, we won't try to dispatch at all.

File: protocol_test.go [hunk 9]
```diff
@@ -97,16 +98,26 @@ func TestLongCallCancel(t *testing.T) {
...
+	resultCh := make(chan result)
+	runInBg(func() error {
 		longResult, err = cli.LongCall(ctx)
-		return err
+		resultCh <- result{longResult, err}
+		longResult, err = cli.LongCallResult(context.Background())
+		resultCh <- result{longResult, err}
+		return nil
 	})
+	time.Sleep(1 * time.Millisecond)
```
Comment (akalin-keybase) [inline protocol_test.go]: May as well just use `runtime.Gosched()` here

File: request.go [hunk 10]
```diff
@@ -51,7 +51,7 @@ func (r *callRequest) LogCompletion(res interface{}, err error) {
 func (r *callRequest) Reply(enc encoder, res interface{}, errArg interface{}) (err error) {
 	select {
 	case <-r.ctx.Done():
-		err = newCanceledError(r.Name(), r.SeqNo())
+		err = r.ctx.Err()
```
Comment (akalin-keybase) [inline request.go]: Since you fix the hang below, you don't need this outer select -- the cancellation will be picked up by EncodeAndWrite or the second select below. So may as simplify it.

File: protocol_test.go (hunk 9 above)
Comment (jzila) [inline protocol_test.go]: Didn't work consistently.

File: dispatch.go (hunk 8 above)
Comment (akalin-keybase) [inline dispatch.go]: That's fine, but now the regression test you added for the cancellation hang bug (TestDispatchCallAfterClose) now tests something else trivial. So if you really want to add this, you'll have to add another test that actually tests what TestDispatchCallAfterClose used to test (i.e., an RPC call makes it past this select, but then Close() is called, and it shouldn't hang).

Also, if you're gonna do this here, you probably should do it for Notify, too. And add similar tests, etc. (I notice there are none for Notify).

Seems better to just keep this PR to just fixing the bug, and not adding any more code paths.

**Comment by akalin-keybase** [comment]
One more pass.

**Comment by jzila** [comment]
Updated

File: codec.go [hunk 11]
```diff
@@ -12,7 +13,8 @@ type decoder interface {
 
 // TODO rename this interface to EncodeAndSend
 type encoder interface {
-	Encode(interface{}) <-chan error
+	EncodeAndWrite(context.Context, interface{}) <-chan error
```
Comment (akalin-keybase) [inline codec.go]: I guess you can remove the TODO above now

File: protocol_test.go (hunk 9 above)
Comment (akalin-keybase) [inline protocol_test.go]: Does this test depend on the sleep working? If so, then that means the test is buggy. :/ You can't rely on the Sleep causing some other particular goroutine to run.

Can you rewrite the test without requiring sleeps?

**Comment by akalin-keybase** [comment]
lgtm after nits and fixing that one test.
